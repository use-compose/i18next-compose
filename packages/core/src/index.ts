export { initI18nConfig, type Createi18nConfigParams } from './init-i18n-config';
export * from './types';
export * from './utils';
