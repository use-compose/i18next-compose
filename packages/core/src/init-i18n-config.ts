import { createInstance, i18n, InitOptions, Resource } from 'i18next';
import FsBackend, { FsBackendOptions } from 'i18next-fs-backend';

export { initI18nConfig };

const i18NConfigOptions: InitOptions = {
  fallbackLng: 'en',
  preload: ['en', 'en'],
  interpolation: {
    escapeValue: false,
  },
  initImmediate: true,
};

export interface Createi18nConfigParams {
  namespace: string;
  fallbackLng: string;
  preload?: string[];
  lng: string;
  supportedLanguages: string[];
  resources?: Resource;
  debug?: boolean;
  initImmediate?: boolean;
  backend?: FsBackendOptions;
}

/* 
 * In case you want to load backend resources, you can use the following example to get the i18n instance:

*/

async function initI18nConfig({
  namespace,
  fallbackLng,
  preload,
  lng,
  supportedLanguages,
  resources,
  initImmediate = true,
  backend,
}: Createi18nConfigParams): Promise<i18n> {
  const initializedI18n = createInstance();

  const i18NConfig: InitOptions = {
    ...i18NConfigOptions,
    ns: namespace,
    fallbackLng,
    preload,
    lng,
    supportedLngs: supportedLanguages,
    resources,
    initImmediate,
  };

  if (backend) {
    initializedI18n.use(FsBackend);
    const fsBackendConfig: FsBackendOptions = {
      loadPath: backend?.loadPath,
      addPath: backend?.addPath,
      ident: backend?.ident,
      parse: backend?.parse,
      stringify: backend?.stringify,
      expirationTime: backend?.expirationTime,
    };
    i18NConfig.backend = fsBackendConfig;
  }

  await initializedI18n.init(i18NConfig);
  return initializedI18n;
}
