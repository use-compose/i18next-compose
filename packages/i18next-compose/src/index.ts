export * from '@use-compose/i18next-core';
export * from './compose-context';
export type * from './types';
