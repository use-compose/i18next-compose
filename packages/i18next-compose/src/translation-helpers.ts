import { cTFunc, I18nFormatterHelper, InputNamespaces } from '@use-compose/i18next-core';
import { Namespace } from 'i18next';

export { composeI18nHelpers };
export type { ComposeI18nHelpers };

interface ComposeI18nHelpers extends I18nFormatterHelper {
  globalNSHelper: cTFunc;
  composeHelper(level2: Namespace): cTFunc;
}

/**
 * Provide helpers and utility function to create own translation helpers
 *
 * @param {I18nApp} i18nextInstance
 * @returns {ComposeI18nHelpers}
 */
function composeI18nHelpers(translationHelper: (level2: Namespace) => cTFunc): ComposeI18nHelpers {
  /**
   * Return helper associated to global namespace
   * @param ns
   * @returns {string}
   *
   * @example
   * {
   *  "namespace": {
   *   "key_from_global_namespace": "This is a global translation"
   *  }
   * }
   *
   * Usage:
   * globalNSHelper('key_from_global_namespace');
   * // returns "This is a global translation"
   */
  const globalNSHelper: cTFunc = (ns: InputNamespaces) => {
    return composeHelper('global')(ns);
  };

  /**
   * Utility to create own translation helper based on level2 namespace
   */
  function composeHelper(level2: string): cTFunc {
    return translationHelper(level2);
  }

  return { translationHelper, globalNSHelper, composeHelper };
}
