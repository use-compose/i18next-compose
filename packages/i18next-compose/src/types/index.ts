export type { I18NextContext } from './context';
