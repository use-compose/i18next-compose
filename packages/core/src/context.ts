import { i18n } from 'i18next';
import { cTFunc } from './types';

export { contextStub };
export type { UseContext };

interface UseContext {
  i18nApp: i18n;
  cT: cTFunc;
  globalNSHelper: cTFunc;
}

function contextStub(): UseContext {
  return {
    i18nApp: {} as i18n,
    cT: () => '',
    globalNSHelper: () => '',
  };
}
