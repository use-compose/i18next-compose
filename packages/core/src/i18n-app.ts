import { useLogger } from '@use-compose/logger';
// import { i18n } from "i18next";

import { createInstance, InitOptions } from 'i18next';
import { Createi18nConfigParams, I18nApp } from './types';

export { configureOptions, createI18nAppInstance };
// interface I18nApp extends i18n {
//   lang:

// TODO: create custom i18n instance that extends i18next i18n

function configureOptions(config: Createi18nConfigParams) {
  // add custom options here
  const baseOptions: InitOptions = {
    fallbackLng: 'en',
    preload: ['en'],
    interpolation: { escapeValue: false },
    initImmediate: true,
  };

  const options = {
    ...baseOptions,
    ...config,
  };

  // if (config.backend) {
  //   options.backend = config.backend;
  // }

  return options;
}

function toInitOptions(cfg: Createi18nConfigParams): InitOptions {
  return {
    ...cfg,
    ns: cfg.namespace,
    supportedLngs: cfg.supportedLanguages,
    backend: cfg.backend,
    resources: cfg.resources,
  };
}

/**
 * Creates an extended i18n instance with custom methods and properties
 * TODO: mandatory?
 *
 * @returns {I18nApp}
 */
function createI18nAppInstance(): I18nApp {
  const { log } = useLogger('createI18nAppInstance');
  // 1. create a base instance
  const base = createInstance();
  // const options = configureOptions(config);
  // keep a reference to the original init
  const originalInit = base.init;

  const extended = Object.assign(base, {
    async initCompose(this: I18nApp, cfg: Createi18nConfigParams): Promise<I18nApp> {
      await originalInit.call(this, toInitOptions(cfg));
      return this;
    },

    // TODO: keep?
    onLangChange(this: I18nApp, callback: (lng: string) => void): () => void {
      this.on('languageChanged', callback);
      // return unsubscribe function
      log(`ðŸ“Ÿ - onLangChange â†’ ${this.language}`);
      return () => {
        this.off('languageChanged', callback);
      };
    },

    missingTranslationHandler(this: I18nApp, key: string) {
      const { language, exists } = this;
      log(`missingTranslationHandler called - ${language}, exists: ${exists(key)}`);
    },
  }) as I18nApp;

  return extended;
}
