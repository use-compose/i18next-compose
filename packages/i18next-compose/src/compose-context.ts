import {
  changeLanguage,
  Createi18nConfigParams,
  cTFunc,
  I18nApp,
  InputNamespaces,
  useCoreContext,
} from '@use-compose/i18next-core';
import { I18NextContext } from './types';

export { composeI18nextContext, i18nextContextStub };

const i18nextContextStub: I18NextContext = {
  i18nApp: {} as I18nApp,
  translationHelper: () => () => '',
  globalNSHelper: () => '',
  composeHelper: () => () => '',
  onLangChange: () => () => {},
  lang: async () => {},
};

/**
 * Creates and returns an i18next context with the provided configuration options.
 *
 * @async
 * @param {Createi18nConfigParams} options
 * @param {{ ssr?: boolean; }} [param0={}]
 * @param {boolean} [param0.ssr=typeof window === 'undefined']
 * @returns {Promise<I18NextContext>}
 */
async function composeI18nextContext(
  options: Createi18nConfigParams,
  { ssr = typeof window === 'undefined' } = {},
): Promise<I18NextContext> {
  if (ssr || !options) {
    return i18nextContextStub;
  }

  const { i18nApp, translationHelper, composeHelper } = await useCoreContext(options);

  /**
   * Return helper associated to global namespace
   * @param ns
   * @returns {string}
   *
   * @example
   * {
   *  "namespace": {
   *   "key_from_global_namespace": "This is a global translation"
   *  }
   * }
   *
   * Usage:
   * globalNSHelper('key_from_global_namespace');
   * // returns "This is a global translation"
   */
  const globalNSHelper: cTFunc = (ns: InputNamespaces) => {
    return composeHelper('global')(ns);
  };

  /**
   * Change the current language.
   *
   * @async
   * @param {?string} [lng]
   * @returns {*}
   */
  async function lang(lng?: string) {
    const target = lng || (options.fallbackLng as string) || 'en';
    if (i18nApp.language !== target) {
      await changeLanguage(i18nApp, target);
    }
  }

  /**
   * Helper/wrapper to provide to specific Frameworks (e.g. Vue, React) to listen to language changes
   * See docs: https://www.i18next.com/overview/api#events
   */
  function onLangChange(cb: (lng: string) => void) {
    i18nApp.on('languageChanged', cb);
    cb(i18nApp.language);
    return function unsubscribe(): void {
      i18nApp.off('languageChanged', cb);
    };
  }

  return {
    i18nApp,
    translationHelper,
    globalNSHelper,
    composeHelper,
    lang,
    onLangChange,
  };
}
