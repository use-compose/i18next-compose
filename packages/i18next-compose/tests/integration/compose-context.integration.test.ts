import { describe, expect, test } from 'vitest';
import { mocki18nConfig } from '../../../core/tests/mocks/i18next-config';
import { composeI18nextContext } from '../../src/compose-context';

describe('Create a translation context and consume it', () => {
  test('Init context and consume hook to get translation', async () => {
    const context = await composeI18nextContext(mocki18nConfig);
    expect(context).toBeDefined();

    const { translationHelper, i18nApp } = context;
    expect(i18nApp).toBeDefined();

    // Access translation function from the context
    const tHelper = translationHelper('compose_translations');
    expect(tHelper).toBeDefined();
    expect(tHelper('test_key')).toEqual('Value of test key2');
    expect(tHelper('test_key')).toMatchSnapshot();
  });
});
