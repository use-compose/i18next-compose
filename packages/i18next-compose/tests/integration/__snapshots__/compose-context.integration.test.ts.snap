// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Create a translation context and consume it > Init context and consume hook to get translation 1`] = `"Value of test key2"`;
